### PR TITLE
MM-45494: Add endpoint for message history

### PR DESCRIPTION
**Pull request by sinansonmez**
#### Summary
This PR creates an endpoint necessary to view history of edited messages and restore an old version with current version.

#### Ticket Link
Fixes https://github.com/mattermost/mattermost-server/issues/20864
JIRA: https://mattermost.atlassian.net/browse/MM-45494

Related Pull Requests

- Has webapp changes (https://github.com/mattermost/mattermost-webapp/pull/11069)

#### Release Note
```release-note
Added new API endpoints GET api/v4/posts/[POST_ID]/edit_history .
```

### CODE REVIEW

**Comment by isacikgoz** [comment]
Hey @sinansonmez I approved tests to run, and it's okay to ask for a review during the implementation (in WIP phase) if you think it would be helpful :) Just let us know!

**Comment by sinansonmez** [comment]
Thanks @isacikgoz for the approval and support

File: app/post.go
Comment (jfrerich) [inline app/post.go:1910]: Is this the correct description for this function?  1/5 this could be removed as I think the function is self-explanatory.

File: store/retrylayer/retrylayer.go
Comment (jfrerich) [inline store/retrylayer/retrylayer.go:6471]: This Pattern looks like it matches the rest of the the file.  Not for this PR, but seems we could refactor and reduce quite a bit of the duplicated code sometime in the future :)

File: app/app_iface.go
Comment (jfrerich) [inline app/app_iface.go:173]: Can you change the comment to match the correct response types?

**Review by jfrerich — CHANGES_REQUESTED** [review]

File: app/post.go
Comment (sinansonmez) [reply app/post.go:1910]: You are right, description is not correct. I deleted it as suggested.

File: store/retrylayer/retrylayer.go
Comment (sinansonmez) [reply store/retrylayer/retrylayer.go:6471]: I agree that there are many duplicated code in this file.

File: app/app_iface.go
Comment (sinansonmez) [reply app/app_iface.go:173]: Updated

**Comment by jfrerich** [comment]
@sinansonmez, can you push your changes? :)

**Comment by sinansonmez** [comment]
Here you go @jfrerich 😄

**Comment by jfrerich** [comment]
Can you add the commits regarding the comments related to the function descriptions? :)

**Comment by sinansonmez** [comment]
@jfrerich [This](https://github.com/mattermost/mattermost-server/pull/20945/commits/75d0b70926d1267f351d3e6dc44b5b54718b8ec2) is the only commit related to function descriptions. Were you expecting changes in more places?

**Comment by jfrerich** [comment]
You're good. I refreshed and see the changes now.  Sorry for the confusion!

**Comment by jfrerich** [comment]
@sinansonmez, it's been a long time since I've been in our server side code.  I'm going to add @agnivade to help out! @agnivade, do you have some guidance for @sinansonmez?

**Comment by agnivade** [comment]
@isacikgoz - Would you be able to assist Sinan here?

File: api4/post.go
Comment (isacikgoz) [inline api4/post.go:495]: I think we should have some permission check here? I don't know too much about the feature but is it allowed to get post history for everyone?

File: api4/post_test.go
Comment (isacikgoz) [inline api4/post_test.go:2969]: Related question with above, if we need a permission check it's better to add cases for permissions. Also a test case if there is no post edits etc.

File: app/post_test.go
Comment (isacikgoz) [inline app/post_test.go:3107]:
```suggestion
		require.Empty(t, edits)
```

**Review by isacikgoz — COMMENTED** [review]
This is a neat feature and I really like how you implemented it 💯 Regarding to CI errors, you can run `make layers` and it would allow for further testing. Apart from that I have a question about if this feature requires any permission check cc @esethna

File: model/client4.go
Comment (isacikgoz) [inline model/client4.go:3975]: BTW, the reason you get error in `api4/post_test.go:2986` is you need to make a `GET` request because the endpoint is defined that way. https://github.com/mattermost/mattermost-server/blob/85488813fe959f600b6c3cdfbfbdb928d3b8bfab/api4/post.go#L25

File: api4/post.go
Comment (sinansonmez) [reply api4/post.go:495]: Good point related to permission checks. 

I believe permission checks should be same as editing post. If a user has permission to edit a post, they should be also able to restore an older version. I am leaving below permission related piece from `patchPost`. I wait for the final decision  from @esethna

https://github.com/mattermost/mattermost-server/blob/0c3704c9ca6ab30f7e643de7ff2ee45e34ae7925/api4/post.go#L816-L826

File: api4/post_test.go
Comment (sinansonmez) [reply api4/post_test.go:2969]: Another good catch on the if trhere is no post edits

File: model/client4.go
Comment (sinansonmez) [reply model/client4.go:3975]: 🤦 I knew that I was missing something basic. Thanks for the support 😃

File: api4/post.go
Comment (esethna) [reply api4/post.go:495]: @sinansonmez @isacikgoz thanks for the question re permission checks.

- I agree that any user that has edit permissions on the post should be able to restore previous versions. 
- However, I think any user (even if they don't have edit permissions on the post) should be able to VIEW the edit history.

Comment (sinansonmez) [reply api4/post.go]: For @isacikgoz information, I copying Eric's latest guidance here https://github.com/mattermost/mattermost-webapp/pull/11069#issuecomment-1256838124

**Comment by sinansonmez** [comment]
@esethna can you please add `Setup Test Cloud Server` label to here. Maybe this will pick up both webapp and server changes

**Comment by sinansonmez** [comment]
unfortunately, this also didnt work. This time test server doesnt get webapp changes

File: api4/post.go
Comment (isacikgoz) [reply api4/post.go]: Thanks for the clarification and the info 💯

**Comment by isacikgoz** [comment]
@sinansonmez The branch names should be identical to build with webapp changes. Can you create a new branch on webapp with the same branch name with server branch?

**Comment by sinansonmez** [comment]
Ohh okay. Good to know for future PRs. I just published the branch [here](https://github.com/sinansonmez/mattermost-webapp/tree/MM-M45494-add-endpoint-for-message-history). I assume we just need to remove and add back `Setup Test Cloud Server`

**Comment by isacikgoz** [comment]
> Should I also create a PR with the same branch, or publishing it remotely is enough? 

You don't need to create a PR, but if you are going to continue working on the webapp PR, you can do so to avoid synchronization work for the feature branch.

**Comment by sinansonmez** [comment]
> You don't need to create a PR, but if you are going to continue working on the webapp PR, you can do so to avoid synchronization work for the feature branch.

Most of the job for webapp is done. Maybe I need 1-2 rounds of iteration would be needed. For this reason, I can syncronize two PRs every time. 

Thanks also for re-adding the label. However I still see, it didnt get webapp changes

**Review by isacikgoz — COMMENTED** [review]
LGTM overall. Only caveat here is that we don't have an index for `originalid` in the Posts table. @agarciamontoro shall we test this with a 12M posts dump? Even though I am not expecting this query to run frequently but even for a single request let's see how long does it take to retrieve results.

**Comment by agarciamontoro** [comment]
Thanks for the heads-up, @isacikgoz :) Yes, we should test this with the 12M posts dump and see how it behaves. We could also try the new community-on-demand feature to play with it :) I'll work on it next week.

What we should also do is adding coverage of the new feature to the [load-test tool](https://github.com/mattermost/mattermost-load-test-ng). I can open a ticket for this. @sinansonmez, would you be open to work on that? No obligation at all to say yes, we can take care of it. But if you want to, I can help you set the whole thing up and answer any questions you have on the implementation :)

**Comment by agarciamontoro** [comment]
Ticket's here: https://mattermost.atlassian.net/browse/MM-45494

Btw, @sinansonmez, this implementation, both server- and webapp-side, is impressive! Thank you for all your work :pray:

**Comment by sinansonmez** [comment]
@agarciamontoro thanks for the compliment. It was really fun to work on new feature implementation end-to-end (from back-end to front-end). Related to `load-test tool`, I don't think, I have the capacity as I have bunch of other things on my plate.

**Comment by agarciamontoro** [comment]
> Related to load-test tool, I don't think, I have the capacity as I have bunch of other things on my plate.

That makes total sense, thanks for the clarity!

**Comment by agarciamontoro** [comment]
Assigning this to myself so I don't forget again about testing performance here.

**Review by agarciamontoro — CHANGES_REQUESTED** [review]
Here are my findings. I tested this in a database with 11841269 (11.8 million) posts on my laptop (i7-11800H 8 cores 16 threads, 32GiB RAM, SSD).

The first time I ran the query in this setup it took ~4s. After some warm-up (i.e., the second and subsequent runs), the query seems to consistently take slightly less than half a second every time it is run. This is explained for the lack of index, of course:

```
mattermost=> EXPLAIN ANALYZE SELECT * FROM Posts WHERE Posts.OriginalId = '9enjkantn7ybignqeagzaesufw' ORDER BY Posts.EditAt DESC;
                                                           QUERY PLAN                                                            
---------------------------------------------------------------------------------------------------------------------------------
 Gather Merge  (cost=575530.38..575568.88 rows=330 width=374) (actual time=457.120..460.766 rows=0 loops=1)
   Workers Planned: 2
   Workers Launched: 2
   ->  Sort  (cost=574530.35..574530.77 rows=165 width=374) (actual time=448.877..448.878 rows=0 loops=3)
         Sort Key: editat DESC
         Sort Method: quicksort  Memory: 25kB
         Worker 0:  Sort Method: quicksort  Memory: 25kB
         Worker 1:  Sort Method: quicksort  Memory: 25kB
         ->  Parallel Seq Scan on posts  (cost=0.00..574524.28 rows=165 width=374) (actual time=448.831..448.832 rows=0 loops=3)
               Filter: ((originalid)::text = '9enjkantn7ybignqeagzaesufw'::text)
               Rows Removed by Filter: 3947090
 Planning Time: 1.768 ms
 JIT:
   Functions: 6
   Options: Inlining true, Optimization true, Expressions true, Deforming true
   Timing: Generation 0.418 ms, Inlining 85.443 ms, Optimization 46.330 ms, Emission 30.755 ms, Total 162.946 ms
 Execution Time: 471.736 ms
(17 rows)

Time: 475,990 ms
```
The obvious solution for this is to create an index on the `OriginalId` column. This takes less than 3 seconds in my setup:

```
mattermost=> CREATE INDEX idx_posts_original_id ON Posts(OriginalId);
CREATE INDEX
Time: 2666,261 ms (00:02,666)
```
After this, the query can do an index scan, making it 10K times faster (0.03ms):

```
mattermost=> EXPLAIN ANALYZE SELECT * FROM Posts WHERE Posts.OriginalId = '9enjkantn7ybignqeagzaesufw' ORDER BY Posts.EditAt DESC;
                                                               QUERY PLAN                                                               
----------------------------------------------------------------------------------------------------------------------------------------
 Sort  (cost=53.38..54.37 rows=395 width=374) (actual time=0.020..0.021 rows=0 loops=1)
   Sort Key: editat DESC
   Sort Method: quicksort  Memory: 25kB
   ->  Index Scan using idx_posts_original_id on posts  (cost=0.43..36.35 rows=395 width=374) (actual time=0.018..0.018 rows=0 loops=1)
         Index Cond: ((originalid)::text = '9enjkantn7ybignqeagzaesufw'::text)
 Planning Time: 0.140 ms
 Execution Time: 0.036 ms
(7 rows)

Time: 0,405 ms
```

This is right after creating the index, a situation where Postgres may have several pages cached. But I've repeated this with other IDs afterwards and I don't get anything more than ~0.15ms (1K times faster, instead of 10K; not bad in any case).

I don't foresee this feature to be used super frequently, but half a second for a query in a DB with no activity (this was the only query running at this time in my setup) seems like a lot, so **I'd recommend adding the index to the DB**.

I will prioritize adding coverage for this feature to the load-test tool once it's merged so we can check its impact in the next performance comparison with v7.6.

**Comment by isacikgoz** [comment]
Great feedback @agarciamontoro and much appreciated! So we think adding an index would be good, but also that comes with a write performance cost right? And the posts table is our busiest table. I'd like to have at least a small discussion in the server guild call. I'll queue this item for the next call. 

Meanwhile @sinansonmez would you be able to add migrations for the index?

**Comment by sinansonmez** [comment]
Thanks for the feedback @isacikgoz  and @agarciamontoro 

After your discussion in the server guild, is there any change? If not, I proceed with adding an index.

**Comment by agarciamontoro** [comment]
> After your discussion in the server guild, is there any change? If not, I proceed with adding an index.

It's on November 10. If you can/want to attend, it's open to everyone, it happens in the [~Developers: Server](https://community.mattermost.com/core/channels/developers-server) channel. If not, I'll let you know of the outcome of the discussion :)

**Comment by sinansonmez** [comment]
@isacikgoz @esethna @jfrerich @agarciamontoro finally test server is working correctly. You can check here https://github.com/mattermost/mattermost-server/pull/21601#issuecomment-1307655892

**Comment by agarciamontoro** [comment]
@sinansonmez, we had the server guild meeting today and decided adding the index will probably be the best way forward, but we want to run load-tests with and without the index first to make sure there's no hit to performance by adding it. I'll come back with the results of these tests next week. Thanks for your infinite patience :pray:

**Comment by sinansonmez** [comment]
Thanks for the info @agarciamontoro I couldn't join, but I watched the recording of the meeting. No worries about waiting, I got still some things to do for the webapp so it is not an issue.

**Comment by agarciamontoro** [comment]
The analysis on Posgres was finished. Although this should be repeated on MySQL, the findings are good enough to unblock this PR once the migration with the index is added. The full analysis is [on Community](https://community.mattermost.com/core/pl/x4a75w5hqir6zrxxo96ut5wgpr), but I'll also copy it here so this discussion is self-contained:

# Posts.OriginalID Index Analysis

We're working on a new feature to show the edit history of any post. As part of that development, we need [a new endpoint to retrieve such history](https://github.com/mattermost/mattermost-server/pull/20945). This endpoint, at the end of the day, comes down to the following query, where `OriginalId` is the ID of the post whose history we want to get back:

``` sql
SELECT * FROM Posts WHERE OriginalId = ?
ORDER BY EditAt DESC
```

This document analyzes the proposal to add an index on the `OriginalId` column to improve this query, focusing on the possible impact on other interactions with the `Posts` table.

## Conclusions

The impact of the index on the `Posts` table is shown to be negligible compared to the improvement on the edit history query. As such, we recommend adding a migration with the index in [the original PR](https://github.com/mattermost/mattermost-server/pull/20945).

This analysis was performed only on PostgreSQL. A similar test for MySQL will be done shortly, but we don't expect big differences, so we can go ahead and unblock the feature development.

## Analysis

### Preliminary analysis

A preliminary analysis showed that the cost of this query may be too high, as shown by the output of this `EXPLAIN ANALYZE` on a database with a `Posts` table with around 12 million posts:

```
mattermost=> EXPLAIN ANALYZE SELECT * FROM Posts WHERE Posts.OriginalId = '9enjkantn7ybignqeagzaesufw' ORDER BY Posts.EditAt DESC;
                                                           QUERY PLAN
---------------------------------------------------------------------------------------------------------------------------------
 Gather Merge  (cost=575530.38..575568.88 rows=330 width=374) (actual time=457.120..460.766 rows=0 loops=1)
   Workers Planned: 2
   Workers Launched: 2
   ->  Sort  (cost=574530.35..574530.77 rows=165 width=374) (actual time=448.877..448.878 rows=0 loops=3)
         Sort Key: editat DESC
         Sort Method: quicksort  Memory: 25kB
         Worker 0:  Sort Method: quicksort  Memory: 25kB
         Worker 1:  Sort Method: quicksort  Memory: 25kB
         ->  Parallel Seq Scan on posts  (cost=0.00..574524.28 rows=165 width=374) (actual time=448.831..448.832 rows=0 loops=3)
               Filter: ((originalid)::text = '9enjkantn7ybignqeagzaesufw'::text)
               Rows Removed by Filter: 3947090
 Planning Time: 1.768 ms
 JIT:
   Functions: 6
   Options: Inlining true, Optimization true, Expressions true, Deforming true
   Timing: Generation 0.418 ms, Inlining 85.443 ms, Optimization 46.330 ms, Emission 30.755 ms, Total 162.946 ms
 Execution Time: 471.736 ms
(17 rows)

Time: 475,990 ms
```

As shown in the `Execution Time`, this query takes almost half a second. Although getting the edit history of a post will not be a frequent action, this time can definitely be improved. The output of the `EXPLAIN ANALYZE` shows the cue to how we can do this: at the core of the plan, the system does a sequential scan on `Posts`; i.e., it iterates through the whole table until it finds the information needed.

This can be easily improved with an index as the following one:

``` sql
CREATE INDEX idx_posts_original_id ON Posts(OriginalId);
```

After this addition, the query planner can now use the index, and the execution time is now much better:

```
mattermost=> EXPLAIN ANALYZE SELECT * FROM Posts WHERE Posts.OriginalId = '9enjkantn7ybignqeagzaesufw' ORDER BY Posts.EditAt DESC;
                                                               QUERY PLAN
----------------------------------------------------------------------------------------------------------------------------------------
 Sort  (cost=53.38..54.37 rows=395 width=374) (actual time=0.020..0.021 rows=0 loops=1)
   Sort Key: editat DESC
   Sort Method: quicksort  Memory: 25kB
   ->  Index Scan using idx_posts_original_id on posts  (cost=0.43..36.35 rows=395 width=374) (actual time=0.018..0.018 rows=0 loops=1)
         Index Cond: ((originalid)::text = '9enjkantn7ybignqeagzaesufw'::text)
 Planning Time: 0.140 ms
 Execution Time: 0.036 ms
(7 rows)

Time: 0,405 ms
```

In subsequent runs of this experiment, we got some other times, but never something above 0.15ms. That is, a \~3000x improvement. Not bad.

But not all are good news. A new index always adds some overhead, since writing to the table now involves updating the index as well. We need to analyze whether the performance hit of the overhead is too big, or if it's negligible enough that we can add the index.

### Index Addition Analysis

The experiment performed simulated a normal user load in the system with 8k users connected over the course of 30 minutes for two cases: a DB without the index, and a DB with the index.

#### Experiment Setup
    
- Mattermost version: v7.4.0
- Database system: PostgreSQL v12
- Dataset used: DB dump from v6.1 with \~12 millions posts
- Number of concurrent users: 8000.
- Time of the test (with all users connected): 30 minutes.
- 2 `c5.2xlarge` server instances.
- 2 `db.r5.2xlarge` database instances.
- Configuration files attached to this post.
    
The experiment used the same deployment (exact same instances) for both setups (with and without index) to minimize variability from AWS instance differences. The first test was performeed without the index, and the second with the index. Before both tests, the database was dropped, the dump from v6.1 was fed into the database and the Mattermost server was restarted. Before restarting the server in the second test, the index was created as shown above.

#### Findings
    
The automatically generated Grafana dashboard can be seen in [this Raintank snapshot](https://snapshots.raintank.io/dashboard/snapshot/LsCoRH8vp6AUww6EEGMdWZfKKAq4xd23) for a birdseye view of both tests. Similarly, we can take a look at the graphs from the automated report comparison, which do not show any significant difference in the metrics observed:

| avg-api-times | p99-store-times |
|---|---|
| avg-store-times | p99-api-times |
| cpu-utilization | goroutines-in-use |
| heap-in-use | number-of-connected-devices |
| rps | stack-in-use |
    
We can make a more dedicated analysis here, though, and look closely into the queries to create and update posts, which are the ones that may be affected by the addition of the index on the `Posts` table.

We can easily query this information in Grafana using the `mattermost_db_store_time` metric. To focus on the specific DB method we want to analyze, we filter by the `method` label. In this case, we're interested in the `PostStore.Save` and `PostStore.Update` methods. For both of them, we'll run the following query (changing `Save` by `Update` where needed for the update case):
    
``` promql
(
    sum(rate(mattermost_db_store_time_sum{method=~"PostStore.Save"}[5m]))
    /
    sum(rate(mattermost_db_store_time_count{method=~"PostStore.Save"}[5m]))
)
-
(
    sum(rate(mattermost_db_store_time_sum{method=~"PostStore.Save"}[5m] offset 7058s))
    /
    sum(rate(mattermost_db_store_time_count{method=~"PostStore.Save"}[5m] offset 7058s))
)
```
    
This looks like a lot, but in short, this is the same as `avg DB query time with index - avg DB query time without index`. That weird offset in the second part of the query is simply the time passed between the beginning of the first test and the beginning of the second, so we can sync both graphs and get a meaningful substraction. Specifying the offset, we simple sample the times from the first test (the offset is into the past). With this Grafana query we get a value over time of the difference of the two setups. In short, if this value is greater than zero, that means that the DB query in the case with the index was slower, and viceversa.
    
If instead of specifying a five-minute window (that's what the `[5m]` means above) we specify a thirty-minute window to cover the duration of the whole test, we can sample the value at the very end, which will contain the average time of the DB method we're querying over the whole test. Now, if we divide that between the same thirty-minute window average of the base case (i.e., the one with no index), then we'll get a single number that will show the percentage of improvement or deterioration of the performance in the corresponding database query.
    
```promql
(
    (
        sum(rate(mattermost_db_store_time_sum{method=~"PostStore.Save"}[30m]))
        /
        sum(rate(mattermost_db_store_time_count{method=~"PostStore.Save"}[30m]))
    )
    -
    (
        sum(rate(mattermost_db_store_time_sum{method=~"PostStore.Save"}[30m] offset 7058s))
        /
        sum(rate(mattermost_db_store_time_count{method=~"PostStore.Save"}[30m] offset 7058s))
    )
)
/
(
    sum(rate(mattermost_db_store_time_sum{method=~"PostStore.Save"}[30m] offset 7058s))
    /
    sum(rate(mattermost_db_store_time_count{method=~"PostStore.Save"}[30m] offset 7058s))
)
```
    
Again, this is super verbose, but it's literally `(avg time with index - avg time without index) / avg time without index`, all within a 30-minute window. If this number is positive, the case with the index was slower; if negative, that means that adding the index was actually good for the performance of that query\! We can of course multiply ths by 100 so we get a percentage, as shown in the screenshot below.
    
Combining these two metrics for both the `Save` and `Update` queries, we get the following dashboard: <https://snapshots.raintank.io/dashboard/snapshot/cSJFwRquhfFq3Y7fYxVDzWijjMi7kveK>
    
screenshot
    
And that gives us our results:
    
- The query to create a post is, in average, 0.2205% slower in the database with the index.
- The query to update a post is, in average, 1.460% faster in the database with the index.
    
Both of these numbers are small enough to discard them as negligible, as the variations between tests (instances in AWS are not always 100% the same) do affect these results. The whole experiment could be repeated to have more confidence in the numbers and assess that indeed it's the setup variations that affect those averages, but there is no need to continue investing much more time on this, since these results are good enough to continue with the initial proposal.
    
That is, we should add the index on `Posts.OriginalId`, whose benefits for the performance of the edit history feature greatly outgrow the impact on the `INSERT` and `UPDATE` queries on the `Posts` table.

**Comment by sinansonmez** [comment]
thanks for the detailed and very well-structured analysis @agarciamontoro. As soon as I finished the changes in front-end, I will handle adding the index to back-end

**Review by isacikgoz — APPROVED** [review]
LGTM, thanks for being patient @sinansonmez 💪

File: db/migrations/postgres/000097_posts_originalid_index.up.sql
Comment (agarciamontoro) [inline db/migrations/postgres/000097_posts_originalid_index.up.sql:1]: I see that you added a commit changing this to lowercase. Is that needed?

**Review by agarciamontoro — APPROVED** [review]
Thank you, @sinansonmez, your work is always excellent :tada:

File: db/migrations/postgres/000097_posts_originalid_index.up.sql
Comment (sinansonmez) [reply db/migrations/postgres/000097_posts_originalid_index.up.sql:1]: Yes, probably not necessary but when I check similar previous PRs (https://github.com/mattermost/mattermost-server/pull/19841 and https://github.com/mattermost/mattermost-server/pull/19799), they used lowercase so I wanted to be consistent

**Comment by stylianosrigas** [comment]
Removing failing testwick, failing with 
```
{"timestamp":"2022-11-30 11:14:02.168 Z","level":"fatal","msg":"Failed to apply database migrations.","error":"driver: postgres, message: failed to execute migration, command: executing_query, originalError: pq: duplicate key value violates unique constraint \"db_migrations_pkey\", query: \n\nINSERT INTO db_migrations (version, name) VALUES (97, 'create_posts_priority')\n","stacktrace":[{"Function":"github.com/mattermost/mattermost-server/v6/store/sqlstore.New","File":"github.com/mattermost/mattermost-server/v6/store/sqlstore/store.go","Line":169},{"Function":"github.com/mattermost/mattermost-server/v6/app/platform.New.func2","File":"github.com/mattermost/mattermost-server/v6/app/platform/service.go","Line":184},{"Function":"github.com/mattermost/mattermost-server/v6/app/platform.New","File":"github.com/mattermost/mattermost-server/v6/app/platform/service.go","Line":232},{"Function":"github.com/mattermost/mattermost-server/v6/app.NewServer","File":"github.com/mattermost/mattermost-server/v6/app/server.go","Line":203},{"Function":"github.com/mattermost/mattermost-server/v6/cmd/mattermost/commands.runServer","File":"github.com/mattermost/mattermost-server/v6/cmd/mattermost/commands/server.go","Line":74},{"Function":"github.com/mattermost/mattermost-server/v6/cmd/mattermost/commands.serverCmdF","File":"github.com/mattermost/mattermost-server/v6/cmd/mattermost/commands/server.go","Line":58},{"Function":"github.com/spf13/cobra.(*Command).execute","File":"github.com/spf13/cobra@v1.6.1/command.go","Line":916},{"Function":"github.com/spf13/cobra.(*Command).ExecuteC","File":"github.com/spf13/cobra@v1.6.1/command.go","Line":1044},{"Function":"github.com/spf13/cobra.(*Command).Execute","File":"github.com/spf13/cobra@v1.6.1/command.go","Line":968},{"Function":"github.com/mattermost/mattermost-server/v6/cmd/mattermost/commands.Run","File":"github.com/mattermost/mattermost-server/v6/cmd/mattermost/commands/root.go","Line":14},{"Function":"main.main","File":"github.com/mattermost/mattermost-server/v6/cmd/mattermost/main.go","Line":23},{"Function":"runtime.main","File":"runtime/proc.go","Line":250},{"Function":"runtime.goexit","File":"runtime/asm_amd64.s","Line":1571}]}
```

**Comment by esethna** [comment]
Added approval on the webapp PR

**Comment by AshishDhama** [comment]
Whoever merge this, please close this https://github.com/mattermost/mattermost-webapp/pull/11555
and https://github.com/mattermost/mattermost-server/pull/21601

**Review by jfrerich — APPROVED** [review]

**Comment by jfrerich** [comment]
https://github.com/mattermost/mattermost-webapp/pull/11069 has been merged.  

These status checks need to be resolved before merging this PR.

image

**Comment by koox00** [comment]
@sinansonmez I think you have to delete the branch of the same name in your webapp's fork.
i.e https://github.com/sinansonmez/mattermost-webapp/tree/MM-M45494-add-endpoint-for-message-history

**Pull request merged**
